### PR TITLE
[CURA-10407] Introduce fractional layer-height support gaps

### PR DESCRIPTION
**Pull request by rburema**
Support gaps (z distance) between the model(s) and the support can be non-multiples of layer-heights now.

### CODE REVIEW

File: include/sliceDataStorage.h [hunk 1]
```diff
@@ -212,6 +212,8 @@ class SupportLayer
     std::vector<SupportInfillPart> support_infill_parts; //!< a list of support infill parts
     Polygons support_bottom; //!< Piece of support below the support and above the model. This must not overlap with any of the support_infill_parts or support_roof.
     Polygons support_roof; //!< Piece of support above the support and below the model. This must not overlap with any of the support_infill_parts or support_bottom.
+    Polygons support_fractional_roof_top; //!< If the support distance is less than a multiple of the layer height,
```
Comment (casperlamboo) [inline include/sliceDataStorage.h:215]:
```suggestion
    Polygons support_fractional_roof_top; //!< If the support distance is not exactly a multiple of the layer height,
```
Bit nitpicky, but the support distance is always less then a multiple of a layer height

```
support distance < layer height * n
```
There always exists an `n` for which the support distance is less the `n * layer height`

**Review by casperlamboo — APPROVED** [review]

**Review by wawanbreton — COMMENTED** [review]
Not much to say, just a few very minor remarks. Didn't get into the detailed behavior of the whole thing, but it seems to be designed as to carefully not break anything existing.

**Comment by wawanbreton** [comment]
I'm afraid the current implementation breaks the supports with a 0 Z-distance (which works properly in 5.5):
image
image

**Comment by wawanbreton** [comment]
Also there is a strange behavior: when I set a 0.2mm Z-distance with a 0.1mm layer height, I get only one empty layer between support and model. On 5.5 I get two empty layers as I would expect.

**Comment by rburema** [comment]
The last two may be related -- there could be an off-by-one error?

**Comment by rburema** [comment]
... haha oh no, that's a bit embarrassing. It does sort of explain all the other bugs as well though.

**Comment by wawanbreton** [comment]
Yeah I think it's all the same problem, and it should be easy to fix.

**Comment by wawanbreton** [comment]
Forget my last comment... When the gap grows, the intermediate layer becomes sparser. That is expected. But when it finally become null-height, then it is printed as if it was full height.

**Comment by rburema** [comment]
I see. Still probably an easy fix, just avoid the 'exactly zero' scenario.

File: include/LayerPlan.h [hunk 2]
```diff
@@ -291,7 +292,7 @@ class LayerPlan : public NoCopy
      * \param p The point to travel to
      * \param path (optional) The travel path to which to add the point \p p
      */
-    GCodePath& addTravel_simple(Point p, GCodePath* path = nullptr);
+    GCodePath& addTravel_simple(const Point p, GCodePath* path = nullptr);
```
[the remark this answers is not in the document]
Comment (rburema) [reply include/LayerPlan.h]: Yes I can :-)

**Comment by wawanbreton** [comment]
Just tested new version : 0mm gap works again, but now it does not take care about the "full" layers for the gap: actual maximum distance between support and model is at most 0.99*layer_height

**Comment by rburema** [comment]
Last commit should fix that, I think. Tested it with support roof on and 0.2 layer height, gaps of 0.4, 0.45, 0.55, 0.6 all gave the expected results (z gaps of; 2 full layers, 2+ layers, 2+ layers, 3 full layers respectively).

**Pull request merged**